contracts/protocol/facets/SellerHandlerFacet.sol

```diff
@@ -94,73 +94,178 @@ contract SellerHandlerFacet is SellerBase {
         } else {
-            // Check that the admin address is unique to one seller id across all roles -- not used or is used by this seller id.
-            uint256 check1 = lookups.sellerIdByOperator[_seller.admin];
-            uint256 check2 = lookups.sellerIdByAdmin[_seller.admin];
-            uint256 check3 = lookups.sellerIdByClerk[_seller.admin];
-            require(
-                (check1 == 0 || check1 == _seller.id) &&
-                    (check2 == 0 || check2 == _seller.id) &&
-                    (check3 == 0 || check3 == _seller.id),
-                SELLER_ADDRESS_MUST_BE_UNIQUE
-            );
+            if (_seller.admin != seller.admin) {
+                require(_seller.admin != address(0), INVALID_ADDRESS);
+                // If admin address exists, admin address owner must approve the update for prevent front-running
+                lookups.sellerPendingUpdates[_seller.id].admin = _seller.admin;
+                needsApproval = true;
+            }
+        }
+
+        if (seller.treasury != _seller.treasury) {
+            require(_seller.treasury != address(0), INVALID_ADDRESS);
+            // Update treasury
+            seller.treasury = _seller.treasury;
+        }
+
+        if (_seller.operator != seller.operator) {
+            require(_seller.operator != address(0), INVALID_ADDRESS);
+            // Operator address owner must approve the update for prevent front-running
+            lookups.sellerPendingUpdates[_seller.id].operator = _seller.operator;
+            needsApproval = true;
+        }
+
+        if (_seller.clerk != seller.clerk) {
+            require(_seller.clerk != address(0), INVALID_ADDRESS);
+            // Clerk address owner must approve the update for prevent front-running
+            lookups.sellerPendingUpdates[_seller.id].clerk = _seller.clerk;
+            needsApproval = true;
+        }
+
+        if (needsApproval) {
+            emit SellerUpdateRolesRequested(_seller.id, _seller, sender);
+        }
+
+        // Notify watchers of state change
+        emit SellerUpdated(_seller.id, seller, _authToken, sender);
+    }
+
+    function approveSellerUpdate(uint256 _sellerId) external {
+        // Cache protocol lookups for reference
+        ProtocolLib.ProtocolLookups storage lookups = protocolLookups();
+        address sender = msgSender();
+
+        // Get seller pending update
+        Seller storage _sellerPendingUpdate = lookups.sellerPendingUpdates[_sellerId];
+
+        require(
+            _sellerPendingUpdate.admin != address(0) ||
+                _sellerPendingUpdate.operator != address(0) ||
+                _sellerPendingUpdate.clerk != address(0),
+            NO_PENDING_SELLER_ROLE_UPDATE
+        );
+
+        sellerPreUpdateChecks(_sellerId, _sellerPendingUpdate.admin, lookups);
+        sellerPreUpdateChecks(_sellerId, _sellerPendingUpdate.operator, lookups);
+        sellerPreUpdateChecks(_sellerId, _sellerPendingUpdate.clerk, lookups);
```

Comment:
What if you do here just one `sellerPreUpdateChecks(_sellerId, sender, lookups);`?

For pending updates where role address does not match the sender, it doesn't even matter if they are unique or not, since they will not be updated.
And for pending updates that matches sender it's sufficient to do the check once, since the address is the same.
It seems to me, that it should be ok.

Reply:
nice catch, done